File: src/algorithms/python/dfs.py
import json

def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))

def dfs(graph, u, visited):
    visited[u] = True
    emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{u}", status="current")
    
    for v, w, edge_id in graph[u]:
        emit("GRAPH_EDGE_HIGHLIGHT", edgeId=edge_id, accepted=True, status="path")
        if not visited[v]:
            dfs(graph, v, visited)
            emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{u}", status="current")
            emit("GRAPH_EDGE_HIGHLIGHT", edgeId=edge_id, accepted=False, status="backtrack")
        else:
            emit("GRAPH_EDGE_HIGHLIGHT", edgeId=edge_id, accepted=False)
            
    emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{u}", status="visited")

if __name__ == "__main__":
    graph = [
        [(1, 1, "e0"), (2, 1, "e1")],
        [(3, 1, "e2"), (4, 1, "e3")],
        [(5, 1, "e4")],
        [],
        [],
        []
    ]
    visited = [False] * len(graph)
    dfs(graph, 0, visited)

File: src/algorithms/cpp/dfs.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

struct Edge { int node; string id; };

void emitNodeHighlight(int node) {
    cout << "{\"type\":\"GRAPH_NODE_HIGHLIGHT\",\"nodeId\":\"n" << node << "\"}" << endl;
}
void emitEdgeHighlight(string edgeId, bool accepted) {
    cout << "{\"type\":\"GRAPH_EDGE_HIGHLIGHT\",\"edgeId\":\"" << edgeId << "\",\"accepted\":" << (accepted ? "true" : "false") << "}" << endl;
}
void emitSystemLog(string message) {
    cout << "{\"type\":\"SYSTEM_LOG\",\"message\":\"" << message << "\",\"level\":\"INFO\"}" << endl;
}

void dfs_recursive(int u, vector<vector<Edge>>& graph, vector<bool>& visited) {
    visited[u] = true;
    emitNodeHighlight(u);
    emitSystemLog("Visiting node n" + to_string(u));

    for (auto& edge : graph[u]) {
        int v = edge.node;
        emitEdgeHighlight(edge.id, true);
        if (!visited[v]) {
            dfs_recursive(v, graph, visited);
        }
    }
}

void emitInit(const vector<vector<Edge>>& graph) {
    cout << "{\"type\":\"INIT\",\"graph\":{\"nodes\":[";
    for (int i = 0; i < (int)graph.size(); i++) {
        if (i > 0) cout << ",";
        cout << "{\"id\":\"n" << i << "\",\"label\":\"" << i << "\",\"x\":" << (i%3)*120-120 << ",\"y\":" << (i/3)*120-60 << "}";
    }
    cout << "],\"edges\":[";
    bool first = true;
    for (int u = 0; u < (int)graph.size(); u++) {
        for (const auto& e : graph[u]) {
            if (!first) cout << ",";
            first = false;
            cout << "{\"id\":\"" << e.id << "\",\"from\":\"n" << u << "\",\"to\":\"n" << e.node << "\"}";
        }
    }
    cout << "],\"isDirected\":true}}" << endl;
}

int main() {
    vector<vector<Edge>> graph(5);
    graph[0] = {{1, "e0"}, {2, "e1"}};
    graph[1] = {{3, "e2"}};
    graph[2] = {{4, "e3"}};
    graph[3] = {{4, "e4"}};
    graph[4] = {};

    emitInit(graph);
    vector<bool> visited(5, false);
    dfs_recursive(0, graph, visited);
    emitSystemLog("DFS traversal complete.");
    return 0;
}
